bayesian_changepoint_detection/online_changepoint_detection.py

```diff
@@ -60,8 +64,100 @@ def update_theta(self, data):
+        :param chunksize: The length of array to pre-allocate.
+            This should be a best guess as to how long your data will be.
+            If you aren't sure or want to save RAM at the expense of speed, leave this as 1
+        """
+        self.chunksize = chunksize
+
+        # We default to the minimum possible degrees of freedom, which is 1 greater than the dimensionality
+        if dof is None:
+            dof = dims + 1
+        # The default mean is all 0s
+        if mu is None:
+            mu = [0]*dims
+        # The default covariance is the identity matrix. The scale is the inverse of that, which is also the identity
+        if scale is None:
+            scale = np.identity(dims)
+
+        # The number of data points we have seen
+        self.n = 0
+
+        # The dimensionality of the dataset (number of variables)
+        self.dims = dims
+
+        # A hack to neaten the code: we store the original parameters here in a list, then immediately expand the array,
+        # extended it to a length of chunksize
+        self.dof = [dof]
+        self.kappa = [kappa]
+        self.mu = [mu]
+        self.scale = [scale]
+
+        self.expand()
+
+    def expand(self):
+        """
+        Increases the length of each array by the chunksize
+        """
+        self.dof = np.concatenate((self.dof, np.empty(self.chunksize)))
+        self.kappa = np.concatenate((self.kappa, np.empty(self.chunksize)))
+        self.mu = np.vstack((self.mu, np.empty((self.chunksize, self.dims))))
+        self.scale = np.vstack((self.scale, np.empty((self.chunksize, self.dims, self.dims))))
+
+    def pdf(self, data):
+        """
+        Returns the probability of the observed data under the current and historical parameters
+        :param data: A 1 x D vector of new data
+        """
+        self.n += 1
+        t_dof = self.dof - self.dims + 1
+        expanded = np.expand_dims((self.kappa * t_dof) / (self.kappa + 1), (1, 2))
+        ret = np.empty(self.n)
+        try:
+            # This can't be vectorised due to https://github.com/scipy/scipy/issues/13450
+            for i, (df, loc, shape) in islice(enumerate(zip(
+                t_dof,
+                self.mu,
+                inv(expanded * self.scale)
+            )), self.n):
+                ret[i] = stats.multivariate_t.pdf(
+                    x=data,
+                    df=df,
+                    loc=loc,
+                    shape=shape
+                )
+        except AttributeError:
+            raise Exception('You need scipy 1.6.0 or greater to use the multivariate t distribution')
+        return ret
+
+    def update_theta(self, data):
+        """
+        Performs a bayesian update on the prior parameters, given data
+        :param data: A 1 x D vector of new data
+        """
+        if self.n >= self.kappa.shape[0]:
+            self.expand()
+        centered = (data - self.mu[self.n - 1])
+
+        self.dof[self.n] = self.dof[self.n - 1] + 1
```

Comment:
Each `param[t]` doesn't depend on `param[t-1]`. Rather, each `param[t, r]` depends on `param[t-1, r-1]`. So I should copy the existing univariate behaviour which updates the entire vector simultaneously as we increase `t`.